Create net utils and move work with sockets there

anosach-luxoft opened the pull request.

anosach-luxoft committed:
Add socket wrapper to project utils

anosach-luxoft committed:
Move wsa_startup helper to utils

anosach-luxoft commented:
@AGaliuzov, @malirod, @pestOO, please review

pestOO commented on src/components/utils/include/utils/net_utils.h:
```diff
@@ -0,0 +1,51 @@
+#include "utils/macro.h"
+
+namespace utils {
+
```
Could we create class instead of methods?
It will be easier to wrap QSocket with Qt porting

malirod replied:
Agree. This looks like almost full socket encapsulation. I think it worth to add some new class e.g. SDLSocket and use it. In this case this class should be somewhere near `tcp_device`

AGaliuzov replied:
And we can use `pImpl` :+1:

pestOO replied:
We also using Sockets in `SocketStreamerAdapter`, `time_tester::Streamer` and `BluetoothSocketConnection`.
So, it possible could be in Utils.

pestOO commented:
Reviewed

AGaliuzov commented:
reviewed

anosach-luxoft commented:
@AGaliuzov, @malirod, @pestOO, pull request has been updated. Please review

pestOO commented on src/components/utils/include/utils/socket.h:
```diff
@@ -0,0 +1,111 @@
+    int level, int optname,
+	  const char* optval, size_t optlen) {
+    if (impl_) {
+      impl_->SetSocketOptions(level, optname, optval, optlen);
```
do not see return here and bellow

pestOO commented on src/components/utils/include/utils/socket.h:
```diff
@@ -0,0 +1,111 @@
+
+class Socket {
+ public:
+  class Impl {
```
Move impl in source file.
We need only forward declaration of it here.

malirod replied:
Agree. Propose not to use inheritance for impl's. write in the header just `class Impl`. And in each cc file (win and posix) put it's own implementation of the impl.

pestOO commented on src/components/utils/src/socket_posix.cc:
```diff
@@ -0,0 +1,117 @@
+
+
+utils::Socket::Socket(int af, int type, int protocol): impl_(NULL) {
+  impl_ = new PosixSocketImpl(af, type, protocol);
```
Could be `; impl_(new...) {}`

pestOO commented on src/components/utils/include/utils/socket.h:
```diff
@@ -0,0 +1,111 @@
+  }
+
+ private:
+  Impl* impl_;
```
Coudl we use reference?

malirod replied:
This will make socket non-copyable. I am not sure that this is good idea.

pestOO replied:
I'm not sure too ;)

pestOO commented on src/components/utils/src/socket_win.cc:
```diff
@@ -0,0 +1,125 @@
+}
+
+utils::WinSocketImpl::WinSocketImpl(int af, int type, int protocol)
+  : socket_(NULL) {
```
seems to be redundant

pestOO commented on src/components/utils/src/socket_win.cc:
```diff
@@ -0,0 +1,125 @@
+  if (INVALID_SOCKET == socket) {
+    socket_ = NULL;
+  } else {
+    socket_ = socket;
```
May be `socket_ = (s != INVALIDE ? s : NULL)`?

malirod replied:
I personally prefer explicit if & else. There can the possibility to add some debug log line. Possibly just for debugging.  If & else suites better in such case.

pestOO replied:
One Debug with a socket value will be enough.

pestOO commented on src/components/utils/src/socket_win.cc:
```diff
@@ -0,0 +1,125 @@
+
+bool utils::WinSocketImpl::BindSocket(
+  const struct sockaddr* addr, size_t addrlen) {
+  if (SOCKET_ERROR == bind(socket_, addr, addrlen)) {
```
May be just return?

anosach-luxoft commented:
@AGaliuzov, @malirod, @pestOO, please review again.

malirod commented:
@anosach-luxoft, I don't see the usage of the Socket class. How to ensure that this code actually works? Another thing is next. In the `tcp_client_listener` there is  the call of the  `ioctlsocket`, which sets the non-blocking mode for the client connection. Possibly this can be done via the `SetOptions`, but this should be double checked.
Usually it's  good to have the possibility to have the way to get the real socket handle from Socket to apply some special API call. E.g. such feature is present in the boost threads. If this is not hard I propose to add such thing. But I don't insist.

AGaliuzov commented:
reviewed

anosach-luxoft committed:
Remove wsa_startup helper from utils

- Remove WSAStartup class from utils
- Call WSAStartup() on SDL start and WSACleanup() on SDL stop

anosach-luxoft commented:
@AGaliuzov, @malirod, pull request has been updated. You'll be able to test this functionality only with my next pull request with audio/video streaming porting where I'll use new wrapper.

malirod commented:
I have no notes. But I'll wait for usage of this class and test both PR's.

anosach-luxoft commented:
@malirod, see 9ed1870

pestOO commented on src/appMain/life_cycle.cc:
```diff
@@ -88,6 +93,15 @@ LifeCycle::LifeCycle()
+
+#ifdef OS_WINDOWS
+  WSAData wsa_data;
+  if (0 != WSAStartup(MAKEWORD(2, 2), &wsa_data)) {
```
Could we move it to appropriate module (WinSockets)?

anosach-luxoft replied:
@pestOO no, because we have to call this function only once

pestOO replied:
``` cplusplus
// file socket.h
class socket {
...
static const bool initatlization_ws;
}

// file socket.cc
bool initialization() {
  // initialization
  ...
  return true;
}
bool socket::initatlization_ws = initialization();
```

anosach-luxoft replied:
Do we really need this?

malirod replied:
I think this is good solution, but not in our case. we have a few places, where we use raw sockets. streamer, tcp transport and possibly somewhere else. So keeping win socket init& cleanup here will not affect other parts. Until we will be sure that all raw sockets are using Socket, better to use explicit WSAStartup& cleanup in the Start/Stop components

anosach-luxoft commented:
@malirod, see cf8bbd0. Also, you can try to check provided functionality with #29

malirod commented:
@anosach-luxoft, rebase this PR

anosach-luxoft commented:
@malirod, rebased

anosach-luxoft committed:
Use std::swap for pointers swapping

anosach-luxoft committed:
Use ssize_t instead of size_t

anosach-luxoft commented:
@malirod, PR has been updated
